pkg/controller/route53hostedzone/zonereconciler.go

```diff
@@ -0,0 +1,336 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package route53hostedzone
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/aws/aws-sdk-go/service/route53"
+	cov1 "github.com/openshift/cluster-operator/pkg/apis/clusteroperator/v1alpha1"
+	lister "github.com/openshift/cluster-operator/pkg/client/listers_generated/clusteroperator/v1alpha1"
+	clustopaws "github.com/openshift/cluster-operator/pkg/clusterapi/aws"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeclientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/aws/aws-sdk-go/aws"
+	controller "github.com/openshift/cluster-operator/pkg/controller"
+
+	coclient "github.com/openshift/cluster-operator/pkg/client/clientset_generated/clientset"
+)
+
+// ZoneReconciler manages getting the desired state, getting the current state and reconciling the two.
+type ZoneReconciler struct {
+	desiredState   *cov1.DNSZone
+	currentState   *route53.HostedZone
+	dnsZonesLister lister.DNSZoneLister
+
+	// logrus logger
+	logger log.FieldLogger
+
+	// Spec stuff
+	dnszone string
+
+	// Kube related fields
+	name      string
+	namespace string
+
+	// AWS related fields
+	coAWSClient clustopaws.Client
+
+	// Kubernetes client
+	kubeClient kubeclientset.Interface
+
+	// Cluster Operator Client for where cluster operator objects are stored.
+	client coclient.Interface
+
+	// Function pointers, useful for dependency injection
+	funcs *functionPointers
```

Comment:
While advocating for removing this below, from what I've seen in go it would be more common to define the function pointers directly on ZoneReconciler and then point them to another function in tests after you create your ZoneReconciler, but before you call Reconcile.